library/std/unstable_state_preparation.qs

```diff
@@ -0,0 +1,446 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Unstable.StatePreparation {
+    open Microsoft.Quantum.Convert;
+    open Microsoft.Quantum.Diagnostics;
+    open Microsoft.Quantum.Arrays;
+    open Microsoft.Quantum.Math;
+
+    /// # Summary
+    /// Given a set of coefficients and a little-endian encoded quantum register,
+    /// prepares a state on that register described by the given coefficients.
+    ///
+    /// # Description
+    /// This operation prepares an arbitrary quantum
+    /// state |𝜓⟩ with coefficients 𝑎ⱼ from
+    /// the n-qubit computational basis state |0...0⟩.
+    ///
+    /// The action of U on the all-zeros state is given by
+    /// $$
+    /// \begin{align}
+    ///     U \ket{0\cdots 0} = \ket{\psi} = \frac{\sum_{j=0}^{2^n-1}\alpha_j \ket{j}}{\sqrt{\sum_{j=0}^{2^n-1}|\alpha_j|^2}}.
+    /// \end{align}
+    /// $$
+    ///
+    /// # Input
+    /// ## coefficients
+    /// Array of up to 2ⁿ real coefficients. The j-th coefficient
+    /// indexes the number state |j⟩ encoded in little-endian format.
+    ///
+    /// ## qubits
+    /// Qubit register encoding number states in little-endian format. This is
+    /// expected to be initialized in the computational basis state |0...0⟩.
+    ///
+    /// # Remarks
+    /// `coefficients` will be normalized and padded with
```

Comment:
I think the pattern of padding arrays if they don't have the expected number of elements can lead to inadvertently using the operation incorrectly. For this reason, I think it is more robust if the operation fails when the number of coefficients is not 2^n.

Reply:
I understand the concern and yes, this is a valid concern. However, it looks like we need to support this functionality. (1) It used to be like that in the old QDK (2) You may want to prepare a sate on only the first five "numbers". Then you won't need to provide coefficients for the other three. (3) It's kind-of consistent with other state preparation routines that aren't ported yet. For example, there's a non-trivial operation that prepares uniform superposition of the first N basis vectors. If N was always 2^n, we wouldn't need that at all, as a simple H and CNOTs would be sufficient.
Given that I'd say we keep the ability to provide only first few coefficients omitting trailing 0s.